internal/service/paymentlinksrv/webhook.go

```diff
@@ -2,33 +2,129 @@ package paymentlinksrv
 
 	paylinkId, err := idValidate(webhook.Transaction.Invoice.PaymentRequestId)
 	if err != nil {
+		aulogging.Logger.Ctx(ctx).Error().Printf("webhook called with invalid reference ID. id=%s", webhook.Transaction.Invoice.PaymentRequestId)
 		return err
 	}
 
 	paylink, err := concardis.Get().QueryPaymentLink(ctx, paylinkId)
 	if err != nil {
+		aulogging.Logger.Ctx(ctx).Error().Printf("can't query payment link from concardis. err=%s", err)
 		return err
 	}
 
-	aulogging.Logger.Ctx(ctx).Info().Printf("paylink id=%d ref=%s status=%s", paylink.ID, paylink.ReferenceID, paylink.Status)
+	if paylink.ReferenceID != webhook.Transaction.Invoice.ReferenceId {
+		// webhook data claimed it was about ref_id A but the paylink is for ref_id B
+		aulogging.Logger.Ctx(ctx).Error().Printf("webhook reference_id mismatch, ref_id in webhook=%s, ref_id in paylink data=%s", webhook.Transaction.Invoice.ReferenceId, paylink.ReferenceID)
+		return WebhookRefIdMismatchErr
+	}
+
+	aulogging.Logger.Ctx(ctx).Info().Printf("webhook call for paylink id=%d ref=%s status=%s", paylink.ID, paylink.ReferenceID, paylink.Status)
+
+	// fetch transaction data from payment service
+	transaction, err := paymentservice.Get().GetTransactionByReferenceId(ctx, paylink.ReferenceID)
+	if err != nil {
+		if err == paymentservice.NotFoundError {
+			// transaction not found in the payment service -> create one.
+			// Note: this should never happen, but we try to recover because someone paid us money for somthing.
+			aulogging.Logger.Ctx(ctx).Error().Printf("webhook reference_id not found in payment service. Creating new transaction. reference_id=%s", paylink.ReferenceID)
+
+			return createTransaction(ctx, paylink)
+		} else {
+			aulogging.Logger.Ctx(ctx).Error().Printf("error fetching transaction from payment service. err=%s", err)
+			return err
+		}
+	}
+
+	// matching transaction was found in the payment service database.
+	// update the values with data from Concardis.
+	return updateTransaction(ctx, paylink, transaction)
+}
+
+func createTransaction(ctx context.Context, paylink concardis.PaymentLinkQueryResponse) error {
+	debitor_id, err := debitorIdFromReferenceID(paylink.ReferenceID)
+	if err != nil {
+		aulogging.Logger.Ctx(ctx).Warn().Printf("webhook couldn't parse debitor_id from reference_id. reference_id=%s", paylink.ReferenceID)
+		// we log a warning, but we continue anyway
+	}
+
+	transaction := paymentservice.Transaction{
+		ID:        paylink.ReferenceID,
+		DebitorID: debitor_id,
+		Type:      paymentservice.Payment,
+		Method:    paymentservice.Credit, // XXX TODO: this is a guess. We use paylink for credit cards only, atm.
+		Amount: paymentservice.Amount{
+			GrossCent: paylink.Amount,
+			Currency:  paylink.Currency,
+			VatRate:   0,
+		},
+		Comment: "Auto-created by cncrd adapter because the reference_id could not be found in the payment service.",
+		Status:  paymentservice.Pending,
+		// omitting EffectiveDate
```

Comment:
Effective Date must be set on payments (Valuta date).

if Cncr does not provide this, use today.